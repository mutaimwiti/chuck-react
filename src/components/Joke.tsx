import React, {MouseEvent} from "react";
import styled from "styled-components";
import {Button, Card} from "react-bootstrap";

import {JokeState} from "../store/reducers/joke";

type JokeProps = {
    category: string,
    joke: JokeState,
    onLoadAnotherJoke(): void,
}

const StyledCard = styled(Card)`
  border: none;
`

const Joke: React.FC<JokeProps> = ({category, joke, onLoadAnotherJoke}) => {
    const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
        event.currentTarget.blur();
        onLoadAnotherJoke();
    };

    return <>
        {
            joke && <StyledCard>
                <Card.Body>
                    <Card.Title>{category}</Card.Title>
                    <Card.Text>{joke.value}</Card.Text>
                    <Button variant="primary" onClick={handleClick}>
                        Another one
                    </Button>
                </Card.Body>
            </StyledCard>
        }
    </>
}

export default Joke;
