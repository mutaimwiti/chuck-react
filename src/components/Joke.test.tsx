import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import Joke, { JokeProps } from './Joke';

describe('Joke', function () {
  let props: JokeProps;

  beforeEach(() => {
    props = {
      category: 'sports',
      joke: {
        icon_url: 'icon/url',
        id: 'someJokeId',
        url: 'joke/url',
        value: 'This is a joke',
      },
      onLoadAnotherJoke: jest.fn(),
      categories: ['sports', 'art', 'food'],
      onSelectCategory: jest.fn(),
    };
  });

  it('should not blow up', () => {
    expect(() => render(<Joke {...props} />)).not.toThrow();
  });

  it('should match snapshot', () => {
    const { container } = render(<Joke {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should display joke', () => {
    const { getByText } = render(<Joke {...props} />);

    expect(getByText('This is a joke')).toBeInTheDocument();
  });

  describe('when user clicks button to load another joke', () => {
    it('should invoke onLoadAnotherJoke', () => {
      const { getByTestId } = render(<Joke {...props} />);

      expect(props.onLoadAnotherJoke).not.toHaveBeenCalled();

      const button = getByTestId('load-another-joke-button');

      fireEvent.click(button);

      expect(props.onLoadAnotherJoke).toHaveBeenCalledTimes(1);
    });
  });

  describe('when user selects a category', () => {
    it('should invoke onSelectCategory with selected category', () => {
      const { getByTestId } = render(<Joke {...props} />);

      expect(props.onSelectCategory).not.toHaveBeenCalled();

      const dropdownButton = getByTestId('category-dropdown').children[0];
      act(() => {
        fireEvent.click(dropdownButton);
      });

      const dropdownItem = getByTestId('category-dropdown-item-art');
      fireEvent.click(dropdownItem);

      expect(props.onSelectCategory).toHaveBeenCalledTimes(1);
      expect(props.onSelectCategory).toHaveBeenCalledWith('art');
    });
  });
});
