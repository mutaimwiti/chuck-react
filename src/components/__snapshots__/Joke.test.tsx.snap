// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke should match snapshot 1`] = `
<div>
  <div
    class="sc-bdfBwQ vWQJH card"
  >
    <div
      class="card-body"
    >
      <div
        class="card-title h5"
      >
        <div
          class="dropdown"
          data-testid="category-dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-primary"
            type="button"
          >
            sports
          </button>
        </div>
      </div>
      <p
        class="card-text"
      >
        This is a joke
      </p>
      <button
        class="btn btn-outline-secondary"
        data-testid="load-another-joke-button"
        type="button"
      >
        Another 
        <b>
          sports
        </b>
         joke
      </button>
    </div>
  </div>
</div>
`;
